modules/environment/default.nix

```diff
@@ -159,6 +159,7 @@ in
     # Use user, default and system profiles.
     environment.profiles = mkMerge [
       (mkOrder 800 [ "$HOME/.nix-profile" ])
+      (mkOrder 800 [ "\${XDG_STATE_HOME:-$HOME/.local/state}/nix/profile" ])
```

Comment:
I was looking at this today and noticed that in https://github.com/NixOS/nixpkgs/pull/260764 this was updated to remove the shell expansion:

https://github.com/NixOS/nixpkgs/blob/3a90e2d5cba72e4de0516fb93a9ac8a11dcb5252/nixos/modules/config/users-groups.nix#L799-L804

Not sure if a similar approach should be used here, at least for consistency?

Reply:
When trying to add these manually to `environment.profiles`, I also noticed that the `${XDG_STATE_HOME}` escape interacts poorly with clo4’s workaround for fish paths https://github.com/LnL7/nix-darwin/issues/122#issuecomment-1659465635. Instead I just used this, hard-coding `"$XDG_STATE_HOME/nix/profile"` without the braces:

```nix
    # FIXME: Needed to address bug where $PATH is not properly set for fish.
    #        Also fixed for home-manager in home/darwin/programs/fish.nix
    #        See: https://github.com/LnL7/nix-darwin/issues/122#issuecomment-1659465635
    programs.fish.loginShellInit =
      let
        # This naive quoting is good enough in this case. There shouldn't be any
        # double quotes in the input string, and it needs to be double quoted in case
        # it contains a space (which is unlikely!)
        dquote = str: "\"${str}\"";

        # FIXME: Workaround for incorrect XDG paths https://github.com/LnL7/nix-darwin/issues/947
        #
        # Nix-darwin should set these here: https://github.com/LnL7/nix-darwin/blob/91010a5613ffd7ee23ee9263213157a1c422b705/modules/environment/default.nix#L159-L163
        # as is set for nixOS: https://github.com/Gerg-L/nixpkgs/blob/6e8bbb279bd525b516e49021e5cb976f0b1f79e4/nixos/modules/config/users-groups.nix#L764-L769
        # In the future we could use `config.environment.profiles` instead of hard-coding these.
        profiles = [
          "$HOME/.nix-profile"
          "$XDG_STATE_HOME/nix/profile"
          "$HOME/.local/state/nix/profile"
          "/etc/profiles/per-user/$USER"
        ];
      in ''
        fish_add_path --move --prepend --path ${lib.concatMapStringsSep " " (path: dquote "${path}/bin") profiles}
        set fish_user_paths $fish_user_paths
      '';
```

Not sure if a better solution to #122 would solve this… I don't know if NixOS’ `environment.paths` breaks fish users - I would have imagined that would be fixed by now.

Reply:
No that won't work as what happens if $XDG_STATE_HOME is not defined.

#122 issue is macOS specific and has to do with Apple's path_helper which is just broken.(It forces things to the end of the path so you cannot override macOS system binaries)

Also I think it is more complex as I use home-manager as well - for fish I just manually set nix paths after all bits of nix and fish have finished playing around.

This fix is for bash and zsh setup, how it works on fish  I have not really tested as as you note it gets confusing.

The issue is that the XDG path if it exists is not on $PATH in current nix-darwin

Reply:
My issues with not finishing is more not understanding how PRs work with the testing.

Reply:
The nixes one looks incorrect to me

```
 environment.profiles = [
      "$HOME/.nix-profile"
      "\${XDG_STATE_HOME}/nix/profile"
      "$HOME/.local/state/nix/profile"
      "/etc/profiles/per-user/$USER"
    ];
```

Why do we have  "$HOME/.local/state/nix/profile" as if using xdc then the second line is sufficient isn't it.
I will remove the conditional XDG_STATE_HOME though as we probably do not want to set it.


@brendanzab What is the difference between "$XDG_STATE_HOME/nix/profile" and  "\${XDG_STATE_HOME}/nix/profile" I think we need a non xdc user to test.

Reply:
Ahh yeah, that sounds like it could be a problem, for people not using XDG.

The issue for me was that fish didn't like the bash-style expansion (i.e. `${XDG_STATE_HOME}` or`${XDG_STATE_HOME:-$HOME/.local/state}`)  in combination with [clo4's workaround](https://github.com/LnL7/nix-darwin/issues/122#issuecomment-1659465635). Without that workaround maybe the path expansion would be handled before fish gets it? I'm not sure how `environment.profiles` gets provided to fish. Sorry if that is confusing… it's becoming sort of a complicated set of workarounds alas.

Reply:
Feel free to ignore me at any rate, I think your original solution might be better… but yeah might break people’s workarounds unexpectedly, hah.

Reply:
> #122 issue is macOS specific and has to do with Apple's path_helper which is just broken.(It forces things to the end of the path so you cannot override macOS system binaries)

Ahh, ended up finding https://gist.github.com/Linerre/f11ad4a6a934dcf01ee8415c9457e7b2, which described the problem better. From what I gather the main issue is that it re-orders the path.

Reply:
As for fish my config does not seem to have a problem with the syntax here.
My nix-darwin config includes
```
 programs = {
      bash.enable = true;
      zsh.enable = true;
      fish = {
        enable = true;
        useBabelfish = true;
        # babelfishPackage = pkgs.babelfish;
        vendor = {
          completions.enable = true;
          config.enable = true;
          functions.enable = true;
        };
        # xonsh.enable = true;
      };
    };
    environment = {
      shells = with pkgs; [
        bash
        zsh
        fish
      ];

      # Install completions for system packages.
      pathsToLink = [
        "/share/bash-completion"
        "/share/fish"
        "/share/zsh"
      ];
    };
```

Reply:
From https://github.com/LnL7/nix-darwin/pull/962#discussion_r1705814524
> Why do we have "$HOME/.local/state/nix/profile" as if using xdc then the second line is sufficient isn't it.

Unfortunately not. XDG specifies that if `XDG_STATE_HOME` isn’t set, then `$HOME/.local/state` should be used. On the NixOS side, it looks like `${XDG_STATE_HOME:-$HOME/.local/state}` interacted poorly with PAM, and so replacing the parameter expansion with two separate paths (the first hopefully not existing if `XDG_STATE_HOME` isn’t set) seems like a reasonable workaround.

I have
```nix
environment.profiles = lib.mkOrder 801 [
  "$XDG_STATE_HOME/nix/profile"
  "$HOME/.local/state/nix/profile"
];
```
in my darwin-configuration.nix, and it is working fine (but I do have everything using XDG, so I can’t test the non-XDG situation).